consul/autopilot.go

```diff
@@ -0,0 +1,298 @@
+		if err := addFirst.Error(); err != nil {
+			return fmt.Errorf("failed to add raft peer: %v", err)
+		}
+		addSecond := b.server.raft.AddVoter(promotions[i+1].ID, promotions[i+1].Address, 0, 0)
+		if err := addSecond.Error(); err != nil {
+			return fmt.Errorf("failed to add raft peer: %v", err)
+		}
+		newServers = true
+	}
+
+	// If we added a new server, trigger a check to remove dead servers
+	if newServers {
+		select {
+		case b.server.autopilotRemoveDeadCh <- struct{}{}:
+		default:
+		}
+	}
+
+	return nil
+}
+
+// queryServerHealth fetches the raft stats for the given server and uses them
+// to update its ServerHealth
+func (s *Server) queryServerHealth(member serf.Member, server *agent.Server,
+	autopilotConf *structs.AutopilotConfig) (*structs.ServerHealth, error) {
+	stats, err := s.getServerStats(server)
+	if err != nil {
+		return nil, fmt.Errorf("error getting raft stats: %s", err)
+	}
+
+	health := &structs.ServerHealth{
+		ID:             server.ID,
+		Name:           server.Name,
+		SerfStatusRaw:  member.Status,
+		SerfStatus:     member.Status.String(),
+		LastContactRaw: -1,
+		LastContact:    stats.LastContact,
+		LastTerm:       stats.LastTerm,
+		LastIndex:      stats.LastIndex,
+	}
+
+	if health.LastContact != "never" {
+		health.LastContactRaw, err = time.ParseDuration(health.LastContact)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing last_contact duration: %s", err)
+		}
+	}
+
+	// Set LastContact to 0 for the leader
+	if s.config.NodeName == member.Name {
+		health.LastContactRaw = 0
+		health.LastContact = "leader"
+	}
+
+	health.Healthy = s.isServerHealthy(health, autopilotConf)
+
+	// If this is a new server or the health changed, reset StableSince
+	lastHealth := s.getServerHealth(server.Addr.String())
+	if lastHealth == nil || lastHealth.Healthy != health.Healthy {
+		health.StableSince = time.Now()
+	} else {
+		health.StableSince = lastHealth.StableSince
+	}
+
+	return health, nil
+}
+
+func (s *Server) getServerHealth(addr string) *structs.ServerHealth {
+	s.autopilotLock.RLock()
+	defer s.autopilotLock.RUnlock()
+	h, ok := s.autopilotHealth[addr]
+	if !ok {
+		return nil
+	}
+	return h
+}
+
+func (s *Server) getServerStats(server *agent.Server) (structs.ServerStats, error) {
+	var args struct{}
+	var reply structs.ServerStats
+	err := s.connPool.RPC(s.config.Datacenter, server.Addr, server.Version, "Status.RaftStats", &args, &reply)
+	return reply, err
+}
+
+// isServerHealthy determines whether the given ServerHealth is healthy
+// based on the current Autopilot config
+func (s *Server) isServerHealthy(health *structs.ServerHealth, autopilotConf *structs.AutopilotConfig) bool {
+	if health.SerfStatusRaw != serf.StatusAlive {
+		return false
+	}
+
+	if health.LastContactRaw > autopilotConf.LastContactThreshold || health.LastContactRaw < 0 {
+		return false
+	}
+
+	lastTerm, _ := strconv.ParseUint(s.raft.Stats()["last_log_term"], 10, 64)
```

Comment:
Let's add an `error` return here and make it bail out if there's bad data in the stats. This'll silently bite us.